### DISCUSSION TITLE
Plans for Integration with Traditional React Component Libraries?

### DISCUSSION THREAD

**Discussion by tsengia** [DioxusLabs/dioxus#1506] [General]
## Aside
Hello, I am new to Dioxus, new to Rust, and new to React as well so forgive me if I'm bringing up something that is already answered/in the works, or if this question makes no sense.

Dioxus seems like a great solution to JavaScript's performance, poor performance, and abysmal tooling situation.
TypeScript was supposed to provide static typing, but it's also spawned other problems (such as [this massive issue](https://github.com/akutruff/typescript-needs-types)). 

## Background
Many useful component libraries are written in JavaScript/TypeScript/JSX/TSX and will continue to be for a long time 😢 
As far as I can tell, there does not seem to be a documented/formal way to utilize existing React component libraries in Dioxus. 

This results in the following types of efforts:
- Creating new component libraries written in pure Rust/Dioxus
    - https://github.com/DioxusLabs/dioxus/issues/1483
    - https://github.com/hiltonm/dioxus-charts
- Manually port existing component libraries to Dioxus
     - https://github.com/DioxusLabs/dioxus/discussions/1114
- Manually setup hooks between React and Dioxus to achieve integration
     - https://github.com/DioxusLabs/dioxus/discussions/251#discussioncomment-2186669

I find these efforts to be redundant and/or painful, but not performing one of these efforts means that Dioxus developers cannot use the wealth of existing component libraries.

## The Question(s)  
1. **Are there plans to provide ways for developers to utilize existing React component libraries (such as Material UI, Next UI, etc.)?**

2. If plans are in place and/or development is underway, could someone point me in the direction to help contribute to the effort?

3. Does utilizing existing React component libraries go against the entire philosophy of Dioxus?

I find enabling easy usage of existing React component libraries to be a good idea considering the massive number of component libraries already exist.

Additionally, I am led to believe this is possible because ["Dioxus shares React's DNA"](https://dioxuslabs.com/blog/templates-diffing).

Brainstorming Some Possible Solutions:
- No solution: Using existing React component libraries goes against Dioxus's philosphy
- Create tool/transpiler to automate porting of React component libraries to Rust
   - Currently this is my favorite idea and might be feasible
- Add feature to Dioxus to automatically setup [the integration steps detailed here](https://github.com/DioxusLabs/dioxus/discussions/251#discussioncomment-2186669)

**Comment by ealmloff** [comment]
> ## Aside
> Hello, I am new to Dioxus, new to Rust, and new to React as well so forgive me if I'm bringing up something that is already answered/in the works, or if this question makes no sense.
> 
> Dioxus seems like a great solution to JavaScript's performance, poor performance, and abysmal tooling situation. TypeScript was supposed to provide static typing, but it's also spawned other problems (such as [this massive issue](https://github.com/akutruff/typescript-needs-types)).
> 
> ## Background
> Many useful component libraries are written in JavaScript/TypeScript/JSX/TSX and will continue to be for a long time 😢 As far as I can tell, there does not seem to be a documented/formal way to utilize existing React component libraries in Dioxus.
> 
> This results in the following types of efforts:
> 
> * Creating new component libraries written in pure Rust/Dioxus
>   
>   * [UI kit #1483](https://github.com/DioxusLabs/dioxus/issues/1483)
>   * https://github.com/hiltonm/dioxus-charts
> * Manually port existing component libraries to Dioxus
>   
>   * [Component library based on Primer CSS #1114](https://github.com/DioxusLabs/dioxus/discussions/1114)
> * Manually setup hooks between React and Dioxus to achieve integration
>   
>   * [How to integrate with react/angular/vue/stentil app? #251 (reply in thread)](https://github.com/DioxusLabs/dioxus/discussions/251#discussioncomment-2186669)
> 
> I find these efforts to be redundant and/or painful, but not performing one of these efforts means that Dioxus developers cannot use the wealth of existing component libraries.
> 
> ## The Question(s)
> 1. **Are there plans to provide ways for developers to utilize existing React component libraries (such as Material UI, Next UI, etc.)?**
> 2. If plans are in place and/or development is underway, could someone point me in the direction to help contribute to the effort?
> 3. Does utilizing existing React component libraries go against the entire philosophy of Dioxus?

Specifically for react components, there is no work under way, but if you export your react component as a web component, you can use it in Dioxus. See [this example](https://github.com/DioxusLabs/dioxus/blob/a2dd926f7fec89c0cce84359a045917685fd0830/examples/web_component.rs#L9).

We could use web components to easily create a UI library for Dioxus. For example we could use [materail-ui](https://material-web.dev/) web components as long as the web components have no listeners (for example this could be used for icons, and images).

> I find enabling easy usage of existing React component libraries to be a good idea considering the massive number of component libraries already exist.
> 
> Additionally, I am led to believe this is possible because ["Dioxus shares React's DNA"](https://dioxuslabs.com/blog/templates-diffing).
> 
> Brainstorming Some Possible Solutions:
> 
> * No solution: Using existing React component libraries goes against Dioxus's philosphy

I think this is partially true. A library to support using React components in Dioxus would be amazing, but I don't see using React components as a solution to building an ecosystem around Dioxus. There are three main reasons for this:

1) Performance

Performance is one of the areas Dioxus currently shines. If we adopted React libraries as the official solution for many of the libraries Dioxus needs, the performance advantage of Dioxus would go away.

- Reduced work diffing: Dioxus never diffs static parts of the HTML which results in overall faster diffing
- Template cloning: Dioxus creates chunks of elements at a time with template cloning
- Prerendered snippets in SSR: Dioxus prerenders chunks of elements and reuses them in SSR rendering mode. This means that for static parts of HTML, SSR rendering is as simple as a lookup table.

2) API compatibility

Another focus of Dioxus is ergonomics. To make Dioxus easier to work with, we break from the react model in several places. because of this the APIs of React and Dioxus are not fully compatible.

Dioxus handles suspense, hooks, web events, and diffing completely differently from react. These APIs are not cleanly compatible and we shouldn't sacrifice the native rust experience to make react bindings easier

3) Hooks/Listeners

In addition to components, react and dioxus libraries need to export and pass around hooks and listeners. For the web renderer, we can support some version of this through wasm-bindgen and use JSON for other renderers. Creating APIs for every type in a react library that work with wasm-bindgen is a significant amount of work, making those types nice to work with in rust is even more difficult.

> * Create tool/transpiler to automate porting of React component libraries to Rust
>   * Currently this is my favorite idea and might be feasible

Transpiring from untyped javascript to rust with the extra restrictions lifetimes add seems extremely difficult/impossible. Typescript/[assembly script](https://www.assemblyscript.org/) would be somewhat easier, but still a monumental task. Transpiling from jsx to rsx would be somewhat easier, but each component would need some manual work.

> * Add feature to Dioxus to automatically setup [the integration steps detailed here](https://github.com/DioxusLabs/dioxus/discussions/251#discussioncomment-2186669)

I think some version of bindings from react or web-components to Dioxus could be supported. Having a library to make creating bindings would be great! We support the simplest version of foreign components with web components today. We could add support for listeners to web components in Dioxus and add a helper to allow you to use some React components in Dioxus. I don't see this ever fully replacing an ecosystem around Dioxus, but some limited version of bindings could supplement it in some places.

**Comment by tsengia** [comment]
Thank you for this detailed reply! You've given me much more clarity about these things.

I also now understand how impossible it would be to transpile JSX/TSX to Dioxus/Rust, especially since the API's are _similar_ to React, but certainly breaking changes.
